### PR TITLE
Refactor TagsDialog to use FactoryFragment

### LINKED ISSUE #8487: [Bug] Regression CustomStudyDialog filtering on tags will result in a crash
I think this #8341 introduced a bug in `CustomStudyDialog`, it removed the setter for the listener (which is great, since we can't relay on it to be there after a configuration change) but it assumed the fact that `getActivity()` would return an activity that does implement the interface.

In the case of `CustomStudyDialog` the interface was implemented on the dialog, not on its hosing activity `StudyOptionsActivity`.

```
2021-04-05 21:48:15.228 22091-22091/com.ichi2.anki E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.ichi2.anki, PID: 22091
    java.lang.ClassCastException: com.ichi2.anki.StudyOptionsActivity cannot be cast to com.ichi2.anki.dialogs.TagsDialog$TagsDialogListener
        at com.ichi2.anki.dialogs.TagsDialog.lambda$onCreateDialog$1$TagsDialog(TagsDialog.java:171)
        at com.ichi2.anki.dialogs.-$$Lambda$TagsDialog$TJXB9icr-xFZpv9Zstg5vm6cF3o.onClick(Unknown Source:2)
        at com.afollestad.materialdialogs.MaterialDialog.onClick(MaterialDialog.java:410)
        ......
        ......
```

###### Reproduction Steps

1. Open custom study dialog
2. select tag filtering
3. have fun in the tag dialog
4. once you click on ok, it crashes

###### Expected Result
no crash

###### Actual Result
crash

### PR DESCRIPTION
**Pull request by TarekkMA**
## Purpose / Description
#8341 introduced a bug in CustomStudyDialog, it removed the setter for the listener (which is great, since we can't rely on it to be there after a configuration change) but it assumed the fact that getActivity() would return an activity that does implement the interface.

In the case of `CustomStudyDialog`, the interface was implemented on the dialog, not on its hosing activity `StudyOptionsActivity`.

## Fixes
Fixes #8487

## Approach
By using `ExtendedFragmentFactory`, it was an easy task.

## How Has This Been Tested?

Manually + Unit Tests

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by TarekkMA** [comment]
I don't know if we should wait on #8322 first or not.

**Comment by TarekkMA** [comment]
added a simple UI Test (Robolectric) for TagsDialog

**Comment by TarekkMA** [comment]
**Note** I will still have to study why when I show dialog using `childFragmentManager` it makes both of them disappear.

**Comment by TarekkMA** [comment]
it seems like showing a `FragmentDialog` from a `FragmentDialog`'s `childFragmentManager` is not possible since the parent will be dismissed to view the child, causing the child to also be dismissed.

in that case, we must use the `parentFragmentManager` instead.
<hr>

Removing `android:configChanges` in `AndroidManafest.xml` is required to test if the factory actually works as intended.

**Comment by TarekkMA** [comment]
After I did some research I found out that the best way to communicate between two `DialogFragment`s is through the new [Fragment Result API](https://developer.android.com/guide/fragments/communicate#fragment-result).

I refactored `TagsDialog` to communicate through an implementation of `TagsDialogListener` or Fragment Result API. 
(both cases have tests in `TagsDialogTest`)

<hr>

## Rules of FragmentFactory
* In Activities the initialization of the factories must be done before `super.onCreate`.
* Dialogs should use `parentFragmnetFactory` when needing to display another dialog.
* In normal Fragments initialization of factories is done using `childFragmentFactory` in the `onAttach` method
* [TODO another PR] `StudyOptionsFragment` should manage `CustomStudyDialogFactory` using its `childFragmentManager`. so the 

<hr>

* `CustomStudyDialog` will communicate with `TagsDialog` using Fragment Result API
* Normal (not dialog) Fragments and Activites will use the normal listener
* `DeckPickerContextMenu` needs to show `CustomStudyDialog`, but the communication is done in `DeckPicker`. so no problems here.

TagsDialog

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialogFactory.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+package com.ichi2.anki.dialogs;
+
+import com.ichi2.utils.ExtendedFragmentFactory;
+
+import androidx.annotation.NonNull;
+import androidx.fragment.app.Fragment;
+
+public class TagsDialogFactory extends ExtendedFragmentFactory {
+
+    final TagsDialogListener mListener;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/TagsDialogFactory.java:25]:
```suggestion
    private final TagsDialogListener mListener;
```

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.java [hunk 2]
```diff
@@ -120,6 +123,18 @@ public CustomStudyDialog withArguments(int id, long did, boolean jumpToReviewer)
     }
 
 
+    @Override
+    public void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        getParentFragmentManager().setFragmentResultListener(TagsDialog.FRAGMENT_RESULT_ON_SELECTED_TAGS_KEY, this,
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.java:129]: Are there better designs that we can consider here?

File: AnkiDroid/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -168,7 +168,6 @@
             android:name="com.ichi2.anki.StudyOptionsActivity"
             android:label="StudyOptions"
             android:exported="false"
-            android:configChanges="keyboardHidden|locale|orientation|screenSize"
```
Comment (david-allison-1) [inline AnkiDroid/src/main/AndroidManifest.xml:171]: Why is this necessary? Commit has `DROPME` in the title

**Review by david-allison-1 — COMMENTED** [review]
Looks good! Few comments

File: AnkiDroid/src/main/AndroidManifest.xml (hunk 3 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/AndroidManifest.xml:171]: This commit is needed to test that the fragment factories are behaving as intended.

with `android:configChanges` the activity will not recreate if there is a configuration change, meaning I couldn't test lifecycle changes.

I will drop this commit when this PR ready for merge. But ultimately we want to remove it, as it's not recommended to have according to the documentation.

I have opened  #8589 to track this.

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.java (hunk 2 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.java:129]: Using constants and the bundle is kinda ugly. but this how the docs do it.

I wanted to do it in a more elegant way, but I thought I wait for a review on using Fragment Result Api as is first.

Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.java:129]: An elegant solution would require either reflection or annotation processor. I think most of the bundle annotation libraries I remember are deprecated, eg IcePick.

Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.java:129]: We could use [`AutoValue`](https://github.com/google/auto) to make the class `Parcelable` without much boilerplate.

```java
@AutoValue
public abstract class OnSelectedTagsFragmentResult implements Parcelable {

    public abstract int option();
    public abstract ArrayList<String> tags();

    public static OnSelectedTagsFragmentResult create(int option, ArrayList<String> tags) {
        return new AutoValue_Task(option, tags);
    }
    
    public static OnSelectedTagsFragmentResult fromBundle() {
       return bundle.getParcelable(this.getClass().getName());
    }
    
    public Bundle toBundle() {
      Bundle b = new Bundle();
      b.putParcelable(this.getClass().getName(), this);
      return b;
    }
}
```

Also if we don't want the extra bells and whistles in AutoValue, we can go with something like [Parceler](https://github.com/johncarl81/parceler)

Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.java:129]: we can also do it via kotlin which will save a lot of boilerplate

https://developer.android.com/kotlin/parcelize

**Comment by mikehardy** [comment]
This looks like it took a conflict from your related work, and this is the one that will close a release-2.15 blocker

**Comment by TarekkMA** [comment]
I'm waiting on #8717 and I will fix this PR to have no conflicts.

**Comment by mikehardy** [comment]
Okay - I thought it was part of the chain of PRs but they have moved about a little and I couldn't see the direct link anymore :-)

File: AnkiDroid/src/main/AndroidManifest.xml (hunk 3 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/AndroidManifest.xml]: removed

**Comment by TarekkMA** [comment]
@david-allison-1 should I leave the FragmentResultAPI handling as is for now? or go with a different design?

**Comment by david-allison-1** [comment]
> @david-allison-1 should I leave the FragmentResultAPI handling as is for now? or go with a different design?

(please, bear in mind that I'm new to this style of Android programming, please aim to reject this idea and say "it's fine as it it")

```java
public interface TagsDialogListener {
    void onSelectedTags(List<String> selectedTags, int option);

    default <T extends Fragment & TagsDialogListener> void registerTagsDialogListener(T self) {
        self.getParentFragmentManager().setFragmentResultListener(TagsDialog.FRAGMENT_RESULT_ON_SELECTED_TAGS_KEY, self,
                (requestKey, bundle) -> {
                    final List<String> selectedTags = bundle.getStringArrayList(TagsDialog.FRAGMENT_RESULT_ON_SELECTED_TAGS__TAGS);
                    final int option = bundle.getInt(TagsDialog.FRAGMENT_RESULT_ON_SELECTED_TAGS__OPTION);
                    self.onSelectedTags(selectedTags, option);
                });
    }
}
```

**Comment by TarekkMA** [comment]
Thank you for the suggestion. Using default methods on the interface is really cool and it will move all fragment result API stuff in one place.

I did move everything to the interface, and it now doesn't depend on `TagsDialog`
```java
public interface TagsDialogListener {

    String ON_SELECTED_TAGS_KEY = "ON_SELECTED_TAGS_KEY";
    String ON_SELECTED_TAGS__TAGS = "TAGS";
    String ON_SELECTED_TAGS__OPTION = "OPTION";

    void onSelectedTags(List<String> selectedTags, int option);


    default <F extends Fragment & TagsDialogListener> void registerFragmentResultReceiver(F fragment) {
        fragment.getParentFragmentManager().setFragmentResultListener(
                ON_SELECTED_TAGS_KEY, fragment,
                (requestKey, bundle) -> {
                    final List<String> selectedTags = bundle.getStringArrayList(ON_SELECTED_TAGS__TAGS);
                    final int option = bundle.getInt(ON_SELECTED_TAGS__OPTION);
                    fragment.onSelectedTags(selectedTags, option);
                }
        );
    }

    static TagsDialogListener createFragmentResultSender(FragmentManager fragmentManager) {
        return new TagsDialogListener() {
            @Override
            public void onSelectedTags(List<String> selectedTags, int option) {
                final Bundle bundle = new Bundle();
                bundle.putStringArrayList(ON_SELECTED_TAGS__TAGS, new ArrayList<>(selectedTags));
                bundle.putInt(ON_SELECTED_TAGS__OPTION, option);
                fragmentManager.setFragmentResult(ON_SELECTED_TAGS_KEY, bundle);
            }
        };
    }
}
```

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialogListener.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.ichi2.anki.dialogs.tags;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialogListener.java:1]: copyright header missing

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialog.java [hunk 5]
```diff
@@ -132,6 +130,10 @@ public void onCreate(Bundle savedInstanceState) {
         setCancelable(true);
     }
 
+    @NonNull
+    private TagsDialogListener getTagsDialogListener() {
+        return mListener != null ? mListener : TagsDialogListener.createFragmentResultSender(getParentFragmentManager());
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialog.java:135]: if mListener was null here, do you want to set it to the one you construct for the null case, then just return mListener?

Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialog.java:135]: yeah I guess that would be nicer, but the the listener will be only used once before closing the dialog

Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialog.java:135]: but I will lose the final on `mListener`, and if I will remove final I might as wee to assign the mListener if null in `onCreate` method.

Should I modify it ?

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialogListener.java (hunk 4 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialogListener.java:16]: fixed

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialog.java (hunk 5 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialog.java:135]: if it's only used once I don't think it matters then

**Review by mikehardy — APPROVED** [review]
LGTM

**Comment by david-allison-1** [comment]
@TarekkMA I didn't know you could have variables in an interface, that's awesome, kudos to you!

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialogListener.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+public interface TagsDialogListener {
+
+    String ON_SELECTED_TAGS_KEY = "ON_SELECTED_TAGS_KEY";
+    String ON_SELECTED_TAGS__TAGS = "TAGS";
+    String ON_SELECTED_TAGS__OPTION = "OPTION";
+
+    void onSelectedTags(List<String> selectedTags, int option);
+
+
+    default <F extends Fragment & TagsDialogListener> void registerFragmentResultReceiver(F fragment) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialogListener.java:35]: Optional: I feel this method name should mention tags

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialog.java [hunk 7]
```diff
@@ -153,14 +177,14 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
...
-                .onPositive((dialog, which) -> ((TagsDialogListener)requireActivity())
-                        .onSelectedTags(mTags.copyOfCheckedTagList(), mSelectedOption));
+                .onPositive((dialog, which) -> getTagsDialogListener().onSelectedTags(mTags.copyOfCheckedTagList(), mSelectedOption));
         mDialog = builder.build();
 
         mDialog.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE);
         return mDialog;
     }
 
+
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialog.java:187]: nit: line change

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialog.java [hunk 8]
```diff
@@ -53,6 +50,8 @@
         CUSTOM_STUDY_TAGS
     }
 
+
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialog.java:53]: nit: spacing

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.java [hunk 9]
```diff
@@ -82,7 +84,6 @@
     private static final int DECK_OPTIONS = 107;
     private static final int MORE_OPTIONS = 108;
 
-
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.java:85]: nit: spacing

File: AnkiDroid/src/main/java/com/ichi2/utils/ExtendedFragmentFactory.java [hunk 10]
```diff
@@ -75,9 +75,17 @@ public void setBaseFactory(@Nullable FragmentFactory baseFactory) {
...
+    /**
+     * Attaches the factory to a fragment manager by setting the current fragment factory as the base factory
+     * and updating the fragment manager with the extended factory
+     */
+    public <F extends ExtendedFragmentFactory> F attachToFragmentManager(@NonNull FragmentManager fragmentManager) {
+        mBaseFactory = fragmentManager.getFragmentFactory();
+        fragmentManager.setFragmentFactory(this);
         return (F) this;
     }
+
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/utils/ExtendedFragmentFactory.java:90]: nit: spacing

**Review by david-allison-1 — COMMENTED** [review]
@mikehardy Only nits, consider this an 'approve' after changes/optional changes aren't done

**Comment by david-allison-1** [comment]
Bumped CI

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialogListener.java (hunk 6 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialogListener.java:35]: I didn't include that in the name because `registerFragmentResultReceiver` should register all callbacks.

so if the interface had 3 methods, `registerFragmentResultReceiver ` would call `setFragmentResultListener` three times one for each method.

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.java (hunk 9 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.java]: fixed

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialog.java (hunk 8 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialog.java]: fixed

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialog.java (hunk 7 above)
Comment (TarekkMA) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/tags/TagsDialog.java]: fixed

**Comment by TarekkMA** [comment]
I think this PR is done

**Comment by mikehardy** [comment]
Boom! What an odyssey @TarekkMA :-), thank you

**Comment by mikehardy** [comment]
Hi there!

Apologies for the delay, and if you're already submitted something to OpenCollective for May 2021 PRs merged, you may feel free to ignore this.

Just a friendly notice that we try to process OpenCollective payments monthly - it's time for May 2021 submissions

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

(I only post one comment per person to avoid spamming you, regardless of the number of PRs merged, but this note applies to all PRs merged in May)

Thanks!
